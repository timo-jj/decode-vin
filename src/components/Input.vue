<template>
  <section class="field">
    <label v-if="label">{{ label }}</label>

    <input id="val" v-if="isVal" :value="inputValue" :placeholder="placeholder" type="text" class="field__input" />

    <input id="model" v-else v-model="input.model" :placeholder="placeholder" type="text" @input="onInput" class="field__input" />
  </section>
</template>

<script lang="ts">
import { defineComponent, PropType } from 'vue'

export default defineComponent({
  name: 'Input',

  props: {
    label: {
      type: String as PropType<string>,
      default: '',
    },
    placeholder: {
      type: String as PropType<string>,
      default: '',
    },
    inputValue: {
      type: String as PropType<string | null>,
      default: null,
    },
    isVal: {
      type: Boolean as PropType<boolean>,
      default: false,
    },
  },

  emits: ['change'],

  data() {
    return {
      input: {
        model: '' as string,
      },
    }
  },

  methods: {
    onInput(): void {
      this.$emit('change', this.input.model)
    },
  },
})
</script>

<!-- script setup syntax
<script setup lang="ts">
import { defineEmits, defineProps, PropType, reactive } from 'vue'

defineProps({
  label: {
    type: String as PropType<string>,
    default: '',
  },
  placeholder: {
    type: String as PropType<string>,
    default: '',
  },
  inputValue: {
    type: String as PropType<string | null>,
    default: null,
  },
  isVal: {
    type: Boolean as PropType<boolean>,
    default: false,
  },
})

const emit = defineEmits<{
  (eventName: 'change', model: string): void
}>()

const input = reactive({
  model: '' as string,
})

const onInput = (): void => {
  emit('change', input.model)
}
</script>
-->

<style lang="scss" scoped>
.field {
  label {
    display: block;
    font-size: 12px;
    color: #a4a4a4;
    font-weight: 700;
  }

  &__input {
    border: 0;
    outline: 0;
    border-bottom: 1px solid #a4a4a4;
    padding: 5px 0 10px;
    font-size: 18px;
    background: transparent;
  }
}
</style>
