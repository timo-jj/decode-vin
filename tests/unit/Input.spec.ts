import { shallowMount } from '@vue/test-utils'

import Input from '@/components/Input.vue'

const wrapper = shallowMount(Input, {
  emits: ['change'],
  propsData: {
    isVal: false,
  },
} as any)

describe('Input', () => {
  describe('when rendering', () => {
    it('it renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('on input', () => {
    it('should emit input event', async () => {
      const input = wrapper.find('#model')
      await input.trigger('input')

      expect(wrapper.emitted('change')).toBeTruthy()
    })
  })

  describe('when isVal prop is true', () => {
    beforeEach(async () => {
      await wrapper.setProps({ isVal: true })
    })

    it('should display value input and hide model input', async () => {
      const valInput = wrapper.find('#val')
      const modelInput = wrapper.find('#model')

      expect(valInput.exists()).toBe(true)
      expect(modelInput.exists()).toBe(false)
    })
  })
})
