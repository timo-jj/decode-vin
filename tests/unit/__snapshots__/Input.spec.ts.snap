// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input when rendering it renders correctly 1`] = `
<section
  class="field"
>
  <!--v-if-->
  <input
    class="field__input"
    id="model"
    placeholder=""
    type="text"
  />
</section>
`;
